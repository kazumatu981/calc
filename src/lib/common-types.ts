/**
 * 字句の型
 */
export type TokenType = 'number' | 'operator' | 'leftParen' | 'rightParen';

/**
 * 切り出した字句
 */
export interface Token {
    type: TokenType;
    value: string;
    position: number;
}
