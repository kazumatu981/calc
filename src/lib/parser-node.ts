import { ParserError } from './errors';
import { Token } from './token';

export type NodeType = 'single' | 'binary' | 'paren';

export class ParserNode {
    nodeType: NodeType;
    constructor(nodeType: NodeType) {
        this.nodeType = nodeType;
    }
}

export class SingleNode extends ParserNode {
    isNegative: boolean;
    value: string;
    constructor(value: string, isNegative: boolean | undefined = false) {
        super('single');
        this.value = value;
        this.isNegative = isNegative;
    }
}

export class BinaryNode extends ParserNode {
    public left: ParserNode;
    public right: ParserNode;
    operator: string;
    constructor(left: ParserNode, right: ParserNode, operator: string) {
        super('binary');
        this.left = left;
        this.right = right;
        this.operator = operator;
    }

    public appendToRight(operator: string, nodeTobeAppended: ParserNode): this {
        this.right = new BinaryNode(this.right, nodeTobeAppended, operator);
        return this;
    }
    public static connectTwoNodes(
        node1: ParserNode | undefined,
        node2: ParserNode,
        operatorToken: Token | undefined,
    ): ParserNode {
        if (node1 === undefined) {
            return node2;
        }
        if (operatorToken === undefined) {
            throw new ParserError('operator-must-be-expected');
        }

        if (node1.nodeType !== 'binary') {
            return new BinaryNode(node1, node2, operatorToken.value);
        } else {
            const rootBinaryNode = node1 as BinaryNode;
            if (operatorToken.value === '+' || operatorToken.value === '-') {
                return new BinaryNode(node1, node2, operatorToken.value);
            } else {
                // 右側ノードに新しいノードを付け加える
                return rootBinaryNode.appendToRight(operatorToken.value, node2);
            }
        }
    }
}

export class ParenNode extends ParserNode {
    public childRoot: ParserNode;
    constructor(childRoot: ParserNode) {
        super('paren');
        this.childRoot = childRoot;
    }
}
