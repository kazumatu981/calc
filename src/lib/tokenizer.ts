import { TokenizerError } from './errors';
import type { Token } from './token';
import { CharacterTestUtil, TokenizeUtil } from './tokenizer-utils';

/**
 * 入力文字列を字句に分割する(非同期版)
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export async function tokenizeAsync(input: string): Promise<Token[]> {
    return new Promise((resolve, reject) => {
        try {
            resolve(tokenize(input));
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * 入力文字列を字句に分割する
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export function tokenize(input: string): Token[] {
    const tokens: Token[] = [];
    let currentIndex = 0;
    while (currentIndex < input.length) {
        // 字句を切り出す
        const token = readToken(input, currentIndex);
        if (token === undefined) {
            // 切り出せなかった場合そのまま終了
            // (末尾がスペースだった場合)
            break;
        }
        // 切り出した結果を配列に追加
        tokens.push(token.token);
        currentIndex = token.next;
    }
    return tokens;
}

/**
 * inputのstartIndexから始まる字句を切り出す。
 * @param input 入力文字列
 * @param startIndex 開始位置
 * @returns 切り切り出した字句と次の開始位置
 */
export function readToken(input: string, startIndex: number): { token: Token; next: number } | undefined {
    let currentIndex = TokenizeUtil.skipSpaces(input, startIndex);

    if (currentIndex >= input.length) {
        // 空白文字のまま末尾まで行ったらトークンは何もなかったとみなす。
        return undefined;
    } else if (CharacterTestUtil.isDigit(input, currentIndex)) {
        // 数字を切り出す
        return TokenizeUtil.readNumberToken(input, currentIndex);
    } else if (CharacterTestUtil.isOperator(input, currentIndex)) {
        // 演算子を切り出す
        return TokenizeUtil.readOperatorToken(input, currentIndex);
    } else if (CharacterTestUtil.isLeftParen(input, currentIndex)) {
        // 左括弧
        return TokenizeUtil.readParanToken(input, 'leftParen', currentIndex);
    } else if (CharacterTestUtil.isRightParen(input, currentIndex)) {
        // 右括弧
        return TokenizeUtil.readParanToken(input, 'rightParen', currentIndex);
    } else {
        // 予期せぬ文字を検出した
        throw new TokenizerError('unknown-character', input.charAt(currentIndex));
    }
}
