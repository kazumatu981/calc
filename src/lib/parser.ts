import { Token } from './token';
import { ParserError } from './errors';
import { ParserNode, SingleNode, BinaryNode, ParenNode } from './parser-node';

type ParseMode = 'normal' | 'paren';

export interface ParserOptions {
    startIndex?: number;
    mode?: ParseMode;
}
export class Parser {
    private _tokens: Token[];
    private _currentIndex: number;
    private _mode: ParseMode;

    public constructor(tokens: Token[], options?: ParserOptions) {
        this._tokens = tokens;
        this._mode = options?.mode ?? 'normal';
        this._currentIndex = options?.startIndex ?? 0;
    }

    public parse(): ParserNode {
        let isFist = true;
        let rootNode: ParserNode | undefined = undefined;
        while (this._isNotEnd()) {
            const { operatorToken, node } = this._readNextOperatorAndNode(isFist);
            if (isFist) {
                isFist = false;
            }
            rootNode = BinaryNode.connectTwoNodes(rootNode, node, operatorToken);
        }
        if (rootNode === undefined) {
            throw new ParserError('operator-must-be-expected');
        }
        return rootNode;
    }

    public get currentToken(): Token {
        return this._safeReadTokenAt(this._currentIndex);
    }

    public get nextToken(): Token {
        return this._safeReadTokenAt(this._currentIndex + 1);
    }

    private _isNotEnd() {
        if (this._mode !== 'paren') {
            return this._currentIndex < this._tokens.length - 1;
        } else {
            return (
                this._currentIndex < this._tokens.length - 1 &&
                this._safeReadTokenAt(this._currentIndex).type !== 'rightParen'
            );
        }
    }
    private _readNextOperatorAndNode(isFirst: boolean = false): {
        operatorToken?: Token;
        node: ParserNode;
    } {
        let operatorToken: Token | undefined = undefined;
        let node: ParserNode;
        if (!isFirst) {
            operatorToken = this._readAsOperatorToken();
        }
        if (
            this.currentToken.type === 'number' ||
            (this.currentToken.type === 'operator' && this.currentToken.value === '-')
        ) {
            node = this._readAsSingleNode();
        } else if (this.currentToken.type === 'leftParen') {
            const childParser = new Parser(this._tokens, {
                startIndex: this._currentIndex + 1,
                mode: 'paren',
            });
            const childRoot = childParser.parse();
            node = new ParenNode(childRoot);
            this._currentIndex = childParser._currentIndex + 1;
        } else {
            throw new ParserError('unknown-character');
        }

        return { operatorToken, node };
    }
    private _readAsOperatorToken(): Token | undefined {
        let operatorToken: Token | undefined = undefined;
        if (this.currentToken.type === 'operator') {
            operatorToken = this.currentToken;
            this._currentIndex++;
        } else {
            throw new ParserError('operator-must-be-expected');
        }
        return operatorToken;
    }
    private _readAsSingleNode(): SingleNode {
        let node: SingleNode;
        if (this.currentToken.type === 'number') {
            node = new SingleNode(this.currentToken.value);
            this._currentIndex++;
        } else if (
            this.currentToken.type === 'operator' &&
            this.currentToken.value === '-' &&
            this.nextToken.type === 'number'
        ) {
            node = new SingleNode(this.nextToken.value, true);
            this._currentIndex += 2;
        } else {
            throw new ParserError('unexpected-token');
        }
        return node;
    }
    private _safeReadTokenAt(index: number): Token {
        if (index < this._tokens.length) {
            return this._tokens[index];
        } else {
            throw new ParserError('unknown-character');
        }
    }
}
