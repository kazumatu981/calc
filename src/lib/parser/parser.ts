import { Token } from '../tokenizer/token';
import { ParserError } from '../common/errors';
import { ParserNode, SingleNode, ParenNode } from './parser-node';

type ParseMode = 'normal' | 'paren';

export interface ParserOptions {
    startIndex?: number;
    mode?: ParseMode;
}
export class Parser {
    private _startIndex: number;
    private _tokens: Token[];
    private _currentIndex: number;
    private _mode: ParseMode;

    public constructor(tokens: Token[], options?: ParserOptions) {
        this._tokens = tokens;
        this._mode = options?.mode ?? 'normal';
        this._startIndex = options?.startIndex ?? 0;
        this._currentIndex = this._startIndex;
    }

    public parse(): ParserNode {
        this._currentIndex = this._startIndex;
        let isFist = true;
        let rootNode: ParserNode | undefined = undefined;
        while (this._isNotEnd()) {
            const { operatorToken, node } = this._readNextOperatorAndNode(isFist);
            if (isFist) {
                isFist = false;
            }
            rootNode = ParserNode.connectTwoNodes(rootNode, node, operatorToken);
        }
        if (rootNode === undefined) {
            throw new ParserError('no-token');
        }
        if (this._mode === 'paren' && !this.currentToken.isRightParen) {
            throw new ParserError('paren-must-be-expected');
        }
        return rootNode;
    }

    public get currentToken(): Token {
        return this._safeReadTokenAt(this._currentIndex);
    }

    public get nextToken(): Token {
        return this._safeReadTokenAt(this._currentIndex + 1);
    }

    private _isNotEnd() {
        if (this._mode !== 'paren') {
            return this._currentIndex < this._tokens.length;
        } else {
            return this._currentIndex < this._tokens.length && !this.currentToken.isRightParen;
        }
    }
    private _readNextOperatorAndNode(isFirst: boolean = false): {
        operatorToken?: Token;
        node: ParserNode;
    } {
        let operatorToken: Token | undefined = undefined;
        let node: ParserNode;
        if (!isFirst) {
            operatorToken = this._readAsOperatorToken();
        }
        if (this.currentToken.isNumber || this.currentToken.isNegativeSign) {
            node = this._readAsSingleNode();
        } else if (this.currentToken.isLeftParen) {
            node = this._readAsParenNode();
        } else {
            throw new ParserError('unexpected-token');
        }

        return { operatorToken, node };
    }
    private _readAsOperatorToken(): Token | undefined {
        let operatorToken: Token | undefined = undefined;
        if (this.currentToken.type === 'operator') {
            operatorToken = this.currentToken;
            this._currentIndex++;
        } else {
            throw new ParserError('operator-must-be-expected');
        }
        return operatorToken;
    }
    private _readAsSingleNode(): SingleNode {
        let node: SingleNode;
        if (this.currentToken.isNumber) {
            node = new SingleNode(this.currentToken.value);
            this._currentIndex++;
        } else if (this.currentToken.isNegativeSign && this.nextToken.isNumber) {
            node = new SingleNode(this.nextToken.value, true);
            this._currentIndex += 2;
        } else {
            throw new ParserError('unexpected-token');
        }
        return node;
    }

    private _readAsParenNode(): ParenNode {
        const childParser = new Parser(this._tokens, {
            startIndex: this._currentIndex + 1,
            mode: 'paren',
        });
        const childRoot = childParser.parse();
        this._currentIndex = childParser._currentIndex + 1;
        return new ParenNode(childRoot);
    }
    private _safeReadTokenAt(index: number): Token {
        if (index < this._tokens.length) {
            return this._tokens[index];
        } else {
            console.error('index out of range');
            throw new ParserError('no-token');
        }
    }
}
