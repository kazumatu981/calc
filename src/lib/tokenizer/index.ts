import { Token, TokenType } from './token';
import { Tokenizer } from './tokenizer';

export { Token };
export type { TokenType };

/**
 * 入力文字列を字句に分割する(非同期版)
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export async function tokenizeAsync(input: string): Promise<Token[]> {
    return new Promise((resolve, reject) => {
        try {
            resolve(tokenize(input));
        } catch (e) {
            reject(e);
        }
    });
}

/**
 * 入力文字列を字句に分割する
 * @param input 入力文字列
 * @returns 切り出した字句
 */
export function tokenize(input: string): Token[] {
    return new Tokenizer(input).tokenize();
}
