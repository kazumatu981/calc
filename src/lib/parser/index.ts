import { Parser } from './parser';
import { ParserNode, SingleNode, BinaryNode, ParenNode, type NodeType } from './parser-node';
import { tokenize, type Token } from '../tokenizer';

export { ParserNode, SingleNode, BinaryNode, ParenNode, NodeType };

export async function parseAsync(tokens: Token[]): Promise<ParserNode>;
export async function parseAsync(expression: string): Promise<ParserNode>;
/**
 * 数式文字列または字句から、構文木を作成する（非同期インタフェース）
 * @param args 数式文字列あるいは字句配列
 * @returns 構文木
 */
export async function parseAsync(args: string | Token[]): Promise<ParserNode> {
    if (typeof args === 'string') {
        return new Promise<ParserNode>((resolve, reject) => {
            try {
                resolve(parse(tokenize(args)));
            } catch (e) {
                reject(e);
            }
        });
    } else {
        const token = args as Token[];
        return new Promise<ParserNode>((resolve, reject) => {
            try {
                resolve(parse(token));
            } catch (e) {
                reject(e);
            }
        });
    }
}

export function parse(tokens: Token[]): ParserNode;
export function parse(expression: string): ParserNode;
/**
 * 数式文字列または字句から、構文木を作成する
 * @param args 数式文字列あるいは字句配列
 * @returns 構文木
 */
export function parse(args: string | Token[]): ParserNode {
    if (typeof args === 'string') {
        return new Parser(tokenize(args)).parse();
    } else {
        const token = args as Token[];
        return new Parser(token).parse();
    }
}
