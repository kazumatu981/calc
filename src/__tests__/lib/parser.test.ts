import { tokenize } from '../../lib/tokenizer';
import { parse, parseAsync } from '../../lib/parser';

describe('リグレッションテスト', () => {
    const testCases = [
        '   123',
        '   123   ',
        '123   + 456',
        '123   * 456',
        '123   - 456',
        '123   / 456',
        '123   + 456*123',
        '-123   -456',
        '-123--456',
        '-123*(-456+23)',
        '(12+32)*(45-23)',
    ];

    describe('parse from string', () => {
        testCases.forEach((testCase) => {
            test(`parse ${testCase}`, () => {
                expect(parse(testCase)).toMatchSnapshot();
            });
        });
    });

    describe('parse from tokens', () => {
        testCases.forEach((testCase) => {
            test(`parse ${testCase}`, () => {
                expect(parse(tokenize(testCase))).toMatchSnapshot();
            });
        });
    });

    describe('parseAsync from string', () => {
        testCases.forEach((testCase) => {
            test(`parseAsync ${testCase}`, async () => {
                await expect(parseAsync(testCase)).resolves.toMatchSnapshot();
            });
        });
    });

    describe('parseAsync from tokens', () => {
        testCases.forEach((testCase) => {
            test(`parseAsync ${testCase}`, async () => {
                await expect(parseAsync(tokenize(testCase))).resolves.toMatchSnapshot();
            });
        });
    });
});
