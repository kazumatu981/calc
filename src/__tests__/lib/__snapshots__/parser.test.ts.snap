// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`リグレッションテスト parse from string parse    123    1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parse from string parse    123 1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parse from string parse (12+32)*(45-23) 1`] = `
BinaryNode {
  "left": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "12",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "32",
      },
    },
    "nodeType": "paren",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "45",
      },
      "nodeType": "binary",
      "operator": "-",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parse from string parse -123   -456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from string parse -123*(-456+23) 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": true,
        "nodeType": "single",
        "value": "456",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parse from string parse -123--456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from string parse 123   * 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from string parse 123   + 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from string parse 123   + 456*123 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": BinaryNode {
    "left": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "456",
    },
    "nodeType": "binary",
    "operator": "*",
    "right": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "123",
    },
  },
}
`;

exports[`リグレッションテスト parse from string parse 123   / 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "/",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from string parse 123   - 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse    123    1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parse from tokens parse    123 1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parse from tokens parse (12+32)*(45-23) 1`] = `
BinaryNode {
  "left": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "12",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "32",
      },
    },
    "nodeType": "paren",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "45",
      },
      "nodeType": "binary",
      "operator": "-",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse -123   -456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse -123*(-456+23) 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": true,
        "nodeType": "single",
        "value": "456",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse -123--456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse 123   * 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse 123   + 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse 123   + 456*123 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": BinaryNode {
    "left": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "456",
    },
    "nodeType": "binary",
    "operator": "*",
    "right": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "123",
    },
  },
}
`;

exports[`リグレッションテスト parse from tokens parse 123   / 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "/",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parse from tokens parse 123   - 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync    123    1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync    123 1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync (12+32)*(45-23) 1`] = `
BinaryNode {
  "left": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "12",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "32",
      },
    },
    "nodeType": "paren",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "45",
      },
      "nodeType": "binary",
      "operator": "-",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync -123   -456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync -123*(-456+23) 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": true,
        "nodeType": "single",
        "value": "456",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync -123--456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync 123   * 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync 123   + 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync 123   + 456*123 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": BinaryNode {
    "left": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "456",
    },
    "nodeType": "binary",
    "operator": "*",
    "right": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "123",
    },
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync 123   / 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "/",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from string parseAsync 123   - 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync    123    1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync    123 1`] = `
SingleNode {
  "isNegative": false,
  "nodeType": "single",
  "value": "123",
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync (12+32)*(45-23) 1`] = `
BinaryNode {
  "left": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "12",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "32",
      },
    },
    "nodeType": "paren",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "45",
      },
      "nodeType": "binary",
      "operator": "-",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync -123   -456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync -123*(-456+23) 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": ParenNode {
    "childRoot": BinaryNode {
      "left": SingleNode {
        "isNegative": true,
        "nodeType": "single",
        "value": "456",
      },
      "nodeType": "binary",
      "operator": "+",
      "right": SingleNode {
        "isNegative": false,
        "nodeType": "single",
        "value": "23",
      },
    },
    "nodeType": "paren",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync -123--456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": true,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync 123   * 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "*",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync 123   + 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync 123   + 456*123 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "+",
  "right": BinaryNode {
    "left": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "456",
    },
    "nodeType": "binary",
    "operator": "*",
    "right": SingleNode {
      "isNegative": false,
      "nodeType": "single",
      "value": "123",
    },
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync 123   / 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "/",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;

exports[`リグレッションテスト parseAsync from tokens parseAsync 123   - 456 1`] = `
BinaryNode {
  "left": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "123",
  },
  "nodeType": "binary",
  "operator": "-",
  "right": SingleNode {
    "isNegative": false,
    "nodeType": "single",
    "value": "456",
  },
}
`;
